### PR TITLE
Add „Karabiner Profile Switcher“ extension

### PR DESCRIPTION
**Pull request by lucaschultz**
## Description

The extension allow switching between profiles of [Karabiner Elements](https://karabiner-elements.pqrs.org) and opening the app. The code is largely based on the Keyboard Layout Switcher extension for which I opened a [pull request](https://github.com/raycast/extensions/pull/993) but does not have any dependencies (binary or otherwise) aside from the Raycast API. It uses the [Karabiner CLI](https://karabiner-elements.pqrs.org/docs/manual/misc/command-line-interface/) to switch between the profiles.

## Screencast

https://user-images.githubusercontent.com/45899273/155848454-3a5dc023-6843-4b49-8930-e1ac193fce13.mp4

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)

### CODE REVIEW

File: extensions/karabiner-profile-switcher/src/index.tsx [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  open,
+  Clipboard,
+} from "@raycast/api";
+import { useEffect, useState } from "react";
+import { ILayout } from "./model/interfaces";
+import { isKarabinerCliAvailable, KarabinerManager } from "./model/KarabinerManager";
+
+function toastErrorOptions(message: string, error: string): Toast.Options {
+  return {
+    title: "An Error Occured",
```
Comment (pernielsentikaer) [inline extensions/karabiner-profile-switcher/src/index.tsx:21]:
```suggestion
    title: "An error occurred",
```

File: extensions/karabiner-profile-switcher/src/index.tsx [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  const [profiles, setProfiles] = useState<ILayout[]>([]);
+
+  useEffect(() => {
+    (async () => {
+      try {
+        setProfiles(await KarabinerManager.getAll());
+      } catch (e) {
+        if (!(await isKarabinerCliAvailable())) {
+          const options: Toast.Options = {
+            ...toastErrorOptions("Karabiner Elements CLI Not Available", `${e}`),
```
Comment (pernielsentikaer) [inline extensions/karabiner-profile-switcher/src/index.tsx:51]:
```suggestion
            ...toastErrorOptions("Karabiner Elements CLI not available", `${e}`),
```

File: extensions/karabiner-profile-switcher/src/index.tsx [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            secondaryAction: {
+              title: "Get Karabiner",
+              onAction: () => open("https://karabiner-elements.pqrs.org"),
+              shortcut: { modifiers: ["cmd"], key: "o" },
+            },
+          };
+          await showToast(options);
+        } else {
+          await showToast(toastErrorOptions("Couldn't Get Profiles", `${e}`));
```
Comment (pernielsentikaer) [inline extensions/karabiner-profile-switcher/src/index.tsx:60]:
```suggestion
          await showToast(toastErrorOptions("Couldn't get profiles", `${e}`));
```

File: extensions/karabiner-profile-switcher/src/index.tsx [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        }
+      }
+    })();
+  }, []);
+
+  return (
+    <List isLoading={profiles.length === 0} searchBarPlaceholder="Search available layout...">
```
Comment (pernielsentikaer) [inline extensions/karabiner-profile-switcher/src/index.tsx:67]:
```suggestion
    <List isLoading={profiles.length === 0} searchBarPlaceholder="Search available layouts...">
```

File: extensions/karabiner-profile-switcher/src/model/KarabinerManager.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+  get active(): boolean {
+    return this.title === KarabinerManager.activeInput;
+  }
+
+  public async activate(): Promise<void> {
+    const result = await exec(`${karabinerCliPath} --select-profile '${this.id}'`);
+    const status = result.stdout.split("\n")[0];
+    if (status.startsWith("[error]")) {
+      throw new Error(`Karabiner Profile "${this.title}" Not Found`);
```
Comment (pernielsentikaer) [inline extensions/karabiner-profile-switcher/src/model/KarabinerManager.ts:32]:
```suggestion
      throw new Error(`Karabiner Profile "${this.title}" not found`);
```

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @lucaschultz 👋 

Awesome extension 🔥 I have some feedback for you:

- When Karabiner Elements CLI is not installed, then it should stop the `loading` and return empty array to the list, so it shows “No Results” – right now it keeps loading. 

I tested it, and it does work as expected, good job 💪 

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

**Comment by lucaschultz** [comment]
@pernielsentikaer Good catch! The`loading` is now interrupted when the `Karabiner Elements CLI not available` error `Toast` is shown.

**Review by pernielsentikaer — APPROVED** [review]
Hi @lucaschultz 👋 

Awesome update 🔥 it works as expected – well done! 💣

**Comment by sxn** [comment]
nice work! While testing, I found an issue when you have multiple profiles with the same name:

image

image

Since this seems to be a limitation of [karabiner_cli](https://karabiner-elements.pqrs.org/docs/manual/misc/command-line-interface/) though & this is an edge case, I'll merge the PR anyway.

🎉

**Pull request merged**